### PR TITLE
fix (jkube-kit/enricher) : GitEnricher should implement `enrich` instead of `create`

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description
Related to #2500 

GitEnricher should be executed in `enrich` phase so that it's invoked during OpenShift build.

+ Move GitEnricher execution logic into `enrich` from `create` method
+ Update GitEnricherTest to use updated method
+ Update GitEnricherTest parameterized test to use the parameter instead of using value from global `@BeforeEach`
+ Add `BuildConfig`, `ImageStream`, `Secret` resources to parameterized method source
+ Update integration tests with git annotations that do a strict comparison

## Type of change

 - [X] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.dev/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] My code follows the style guidelines of this project
 - [X] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [X] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [X] I have performed a self-review of my code
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [X] I have added tests that prove my fix is effective or that my feature works
 - [X] New and existing unit tests pass locally with my changes
 - [ ] I tested my code in Kubernetes
 - [X] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
### Eclipse JKube [CI Report](https://github.com/jkubeio/ci)

Started new GH workflow run for https://github.com/eclipse/jkube/pull/2562 (_2024-01-16T14:38:53Z_)

:gear: [JKube E2E Tests (7543190289)](https://github.com/jkubeio/ci/actions/runs/7543190289) :hourglass_flowing_sand:
- :hourglass_flowing_sand: [Bootstrap CI test run (#2562)](https://github.com/jkubeio/ci/actions/runs/7543190289/job/20533629558)

File: jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/GitEnricherTest.java [hunk 1]
```diff
@@ -191,24 +194,29 @@ void setUp() throws Exception {
...
+              arguments(new DeploymentBuilder().withNewMetadata().endMetadata().build()),
+              arguments(new DeploymentConfigBuilder().withNewMetadata().endMetadata().build()),
+              arguments(new ReplicaSetBuilder().withNewMetadata().endMetadata().build()),
+              arguments(new ReplicationControllerBuilder().withNewMetadata().endMetadata().build()),
+              arguments(new DaemonSetBuilder().withNewMetadata().endMetadata().build()),
+              arguments(new StatefulSetBuilder().withNewMetadata().endMetadata().build()),
+              arguments(new JobBuilder().withNewMetadata().endMetadata().build()),
+              arguments(new SecretBuilder().withNewMetadata().endMetadata().build()),
+              arguments(new ImageStreamBuilder().withNewMetadata().endMetadata().build()),
+              arguments(new BuildConfigBuilder().withNewMetadata().endMetadata().build())
```
Comment (manusa) [inline jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/GitEnricherTest.java:209]: Note that due to the changes in GitEnricher, now you need to create the ObjectMetaBuilder entry.

I'm not sure that everytime the resources reach the enricher they will contain an ObjectMetaBuilder instance.

Are you sure about this? (consider also fragments and other corner cases)

Comment (rohanKanojia) [reply jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/GitEnricherTest.java:209]: I'm not sure. I checked with a fragment without metadata but I noticed we always seem to initialize metadata while reading fragment into `HasMetadata`:
https://github.com/eclipse/jkube/blob/e101e93ed4bfad99ce9549d85aa968225326839e/jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/util/KubernetesResourceFragments.java#L138-L139

Comment (rohanKanojia) [reply jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/GitEnricherTest.java:209]: Using `ObjectMetadataBuilder` also adds git labels in `.metadata` of Pod Template, not sure whether we want to have git annotations there.

Comment (rohanKanojia) [reply jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/GitEnricherTest.java:209]: In the case of resources generated via opinionated defaults, I checked various handlers in `org.eclipse.jkube.kit.enricher.handler` package and they all seem to be generating a non null ObjectMeta always:
https://github.com/eclipse/jkube/blob/e101e93ed4bfad99ce9549d85aa968225326839e/jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/DeploymentHandler.java#L42

Once resources go through enricher's `create` phase, `jkube-metadata` is the first enricher in `enrich` phase where various visitors try to create metadata if it doesn't exist:
https://github.com/eclipse/jkube/blob/e101e93ed4bfad99ce9549d85aa968225326839e/jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/api/visitor/MetadataVisitor.java#L134

**Review by manusa — APPROVED** [review]
